Avoid duplicate Access-Control-Allow-Origin headers

TimPickin opened the pull request:
I had the same issue as Steffeng87 (issue #4). This new code checks first if there is a header with the key "Access-Control-Allow-Origin" and only if it does not exist, does it add the header. This avoids the problem of duplicate headers.

TimPickin committed:
Avoid duplicate Access-Control-Allow-Origin headers

I had the same issue as Steffeng87 (issue #4). This new code checks first if there is a header with the key "Access-Control-Allow-Origin" and only if it does not exist, does it add the header. This avoids the problem of duplicate headers.

TimPickin commented:
I forgot the using statement:

using System.Linq;

TimPickin committed:
forgot using statement for linq

mattbrailsford commented:
I'm wondering, if there is already a header set, should we validate it against the configured origin as this is just bypassing the configuration so not sure if that will cause problems / unexpected behavior.

Would you ever have a global Allowed-Origin set that wouldn't match the configuration? if so, I'd say validate it and if it doesn't match, throw an exception?

TimPickin commented:
Hi Matt,
Yes, that is true, it could be unexpected behaviour. 
In my case I installed the nuget package Microsoft.Owin.Cors and then in the app.UseCors(CorsOptions.AllowAll). This solves a problem I was having where the OPTIONS verb was not recognized and the preflight checks failed.
The header is set to * from this setup, so it would match the default AuthU header, so then no exception would be thrown.

TimPickin committed:
Throw error if origin header doesn't match

It checks if there is a header with the key Access-Control-Allow-Origin. If there is, it compares this to the value which is set in the UAuth configuration. It this value does not match, then it throws an exception.

mattbrailsford commented on src/Our.Umbraco.AuthU/Web/Controllers/OAuthController.cs:
```diff
@@ -76,6 +77,25 @@ protected void ProcessClient(OAuthTokenRequest request)
+
+			if (!accessControlHeader.Equals(allowedOrigin, StringComparison.OrdinalIgnoreCase))
+			{
+				throw new Exception($"There is currently a header set with the key Access-Control-Allow-Origin, but the value: {accessControlHeader} differs from the OAuth configured value: {allowedOrigin}");
```
Could we look at throwing a `OAuthResponseException` instead of `Exception` so it's something that can be read from the app? I'd probably use the `HttpStatusCode.InternalServerError` status code and have a msg body like:

    new { invalid_allowed_origin = "INSERT_MSG_HERE" }

This way it's just consistent with how the rest of the API throws errors.

mattbrailsford commented on src/Our.Umbraco.AuthU/Web/Controllers/OAuthController.cs:
```diff
@@ -76,6 +77,25 @@ protected void ProcessClient(OAuthTokenRequest request)
+		}
+		else
+		{
+			HttpContext.Current.Response.Headers.Add("Access-Control-Allow-Origin", Client != null ? Client.AllowedOrigin : Context.Options.AllowedOrigin);
```
Use the `allowedOrigin` variable declared on line 83 instead of doing a null check again.

mattbrailsford commented on src/Our.Umbraco.AuthU/Web/Controllers/OAuthController.cs:
```diff
@@ -76,6 +77,25 @@ protected void ProcessClient(OAuthTokenRequest request)
             }
         }
+	
+	protected void SetOriginHeader()
```
Pedantic, but I'd say rename this to `SetAllowedOriginHeader` for clarity

TimPickin committed:
use OAuthResponseException

I now am using the OAuthResponseException to be consistent with the rest of the code.
Also made the Access-Control-Allow-Origin header value a variable to reduce magic strings.

TimPickin committed:
Renamed method for clarity

mattbrailsford commented:
Perfect!

TimPickin commented:
Hi Matt, thanks for merging this pull request, when do you plan to publish the new version to NuGet?

TimPickin commented:
Hi, just wondering what the status is with pushing this to NuGet?

mattbrailsford commented:
Sorry, just a bit busy atm, will aim for next week when things are a bit quieter
